runner/runner.go

```diff
@@ -0,0 +1,379 @@
+		}
+
+		events, err := getSiteEvents(site.URL)
+		if err != nil {
+			time.Sleep(getEventsBreak)
+			continue
+		}
+
+		for _, event := range events {
+			event.URL = site.URL
+			queue <- event
+		}
+
+		time.Sleep(getEventsBreak)
+	}
+}
+
+func getSiteEvents(site string) ([]event, error) {
+	raw, err := runWpCliCmd([]string{"cron-control", "orchestrate", "runner-only", "list-due-batch", fmt.Sprintf("--url=%s", site), "--format=json"})
+	if err != nil {
+		return make([]event, 0), err
+	}
+
+	siteEvents := make([]event, 0)
+	if err = json.Unmarshal([]byte(raw), &siteEvents); err != nil {
+		if debug {
+			logger.Println(fmt.Sprintf("%+v\n", err))
+		}
+
+		return make([]event, 0), err
+	}
+
+	return siteEvents, nil
+}
+
+func runEvents(workerID int, events <-chan event) {
+	for event := range events {
+		if now := time.Now(); event.Timestamp > int(now.Unix()) {
+			if debug {
+				logger.Printf("runEvents-%d skipping premature job %d|%s|%s for %s", workerID, event.Timestamp, event.Action, event.Instance, event.URL)
+			}
+
+			continue
+		}
+
+		subcommand := []string{"cron-control", "orchestrate", "runner-only", "run", fmt.Sprintf("--timestamp=%d", event.Timestamp), fmt.Sprintf("--action=%s", event.Action), fmt.Sprintf("--instance=%s", event.Instance), fmt.Sprintf("--url=%s", event.URL)}
+
+		_, err := runWpCliCmd(subcommand)
+
+		if err == nil {
+			if heartbeatInt > 0 {
+				atomic.AddUint64(&eventRunSuccessCount, 1)
+			}
+
+			if debug {
+				logger.Printf("runEvents-%d finished job %d|%s|%s for %s", workerID, event.Timestamp, event.Action, event.Instance, event.URL)
+			}
+		} else if heartbeatInt > 0 {
+			atomic.AddUint64(&eventRunErrCount, 1)
+		}
+
+		time.Sleep(runEventsBreak)
+	}
+}
+
+func runWpCliCmd(subcommand []string) (string, error) {
+	subcommand = append(subcommand, "--allow-root", "--quiet", fmt.Sprintf("--path=%s", wpPath))
+	if wpNetwork > 0 {
+		subcommand = append(subcommand, fmt.Sprintf("--network=%d", wpNetwork))
+	}
+
+	wpCli := exec.Command(wpCliPath, subcommand...)
+	wpOut, err := wpCli.CombinedOutput()
+	wpOutStr := string(wpOut)
+
+	if err != nil {
+		if debug {
+			logger.Printf("%s - %s", err, wpOutStr)
+			logger.Println(fmt.Sprintf("%+v\n", subcommand))
+		}
+
+		return wpOutStr, err
+	}
+
+	return wpOutStr, nil
+}
+
+func setUpLogger() {
+	logOpts := log.Ldate | log.Ltime | log.LUTC | log.Lshortfile
+
+	if logDest == "os.Stdout" {
+		logger = log.New(os.Stdout, "DEBUG: ", logOpts)
+	} else {
+		path, err := filepath.Abs(logDest)
+		if err != nil {
+			logger.Fatal(err)
+		}
+
+		logFile, err := os.OpenFile(path, os.O_WRONLY|os.O_CREATE|os.O_APPEND, 0644)
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		logger = log.New(logFile, "", logOpts)
+	}
+}
+
+func validatePath(path *string) {
+	if len(*path) > 1 {
+		var err error
+		*path, err = filepath.Abs(*path)
+
+		if err != nil {
+			fmt.Printf("Error: %s", err.Error())
+			os.Exit(3)
+		}
+
+		if _, err = os.Stat(*path); os.IsNotExist(err) {
+			usage()
```

Comment:
Definitely, this needs improvement. #124 is open, but really should be resolved via this PR.